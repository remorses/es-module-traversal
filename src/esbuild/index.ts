export { traverseWithEsbuild, metaToTraversalResult } from './traverseEsbuild'
