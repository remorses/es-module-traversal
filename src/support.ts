import path from 'path'
import resolve from 'resolve'
import { promises as fsp } from 'fs'
import slash from 'slash'
import { JS_EXTENSIONS, PACKAGE_NAME } from './constants'

export const queryRE = /\?.*$/
export const hashRE = /#.*$/

export const cleanUrl = (url: string) => {
    return url.replace(hashRE, '').replace(queryRE, '')
}

export const debug = require('debug')(PACKAGE_NAME)

export let isRunningWithYarnPnp: boolean
try {
    isRunningWithYarnPnp = Boolean(require('pnpapi'))
} catch {}

export const sleep = (t) => new Promise((res) => setTimeout(res, t))

export async function defaultReadFile(filePath: string): Promise<string> {
    return await (await fsp.readFile(filePath)).toString()
}

export const defaultResolver = (root: string, id: string) => {
    try {
        return (
            resolve.sync(id, {
                basedir: root,
                extensions: [...JS_EXTENSIONS],
                // necessary to work with pnpm
                preserveSymlinks: isRunningWithYarnPnp || false,
            }) || ''
        )
    } catch (e) {
        console.error(`WARN: cannot resolve '${id}' from '${root}'`)
        return ''
    }
}

// this map has same signature as batchedPromiseAll in case i want to refactor and make resolver asynchronous
// export const map = <T, Z>(x: T[], func: (x: T) => Z, _n?: number): Z[] => {
//     return x.map(func)
// }

export function isRelative(x: string) {
    x = cleanUrl(x)
    return x.startsWith('.') || x.startsWith('/')
}

export function isJsModule(x: string) {
    const ext = path.posix.extname(cleanUrl(x))
    return !ext || JS_EXTENSIONS.has(ext)
}

export function flatten<T>(arr: T[][]): T[] {
    return arr.reduce(function (flat, toFlatten) {
        return flat.concat(
            Array.isArray(toFlatten) ? flatten(toFlatten as any) : toFlatten,
        )
    }, [])
}
