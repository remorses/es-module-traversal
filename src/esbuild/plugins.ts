import { Plugin } from 'esbuild'
import builtins from 'builtin-modules'

// resolve node builtins

// mark node modules as external

export function StopTraversingPlugin({ stopTraversing }): Plugin {
    return {
        name: 'stop-traversing',
        setup: function setup({ onLoad, onResolve }) {
            onResolve({ filter: /.*/ }, (args) => {
                console.log({ args })
                const external = stopTraversing(
                    args.resolveDir, // TODO replace resolveDir with the complete path to file
                    args.importer,
                )
                return {
                    external,
                }
            })
        },
    }
}

export function CustomResolverPlugin({ resolver }): Plugin {
    const builtinsSet = new Set(builtins)
    return {
        name: 'stop-traversing',
        setup: function setup({ onLoad, onResolve }) {
            onResolve({ filter: /.*/ }, (args) => {
                if (builtinsSet.has(args.path)) {
                    return {
                        external: true,
                    }
                }
                const resolved = resolver(args.importer, args.path)

                // console.log({ resolved })
                return {
                    external: !resolved,
                    path: resolved,
                }
            })
        },
    }
}
