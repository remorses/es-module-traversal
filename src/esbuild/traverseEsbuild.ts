import {
    NodeResolvePlugin
} from '@esbuild-plugins/all'
import deepmerge from 'deepmerge'
import { build, BuildOptions, Metadata, Plugin } from 'esbuild'
import { promises as fsp } from 'fs'
import fsx from 'fs-extra'
import os from 'os'
import path from 'path'
import slash from 'slash'
import { JS_EXTENSIONS } from '../constants'
import {
    defaultResolver,
    flatten,
    isRunningWithYarnPnp,
    unique
} from '../support'
import { TraversalResultType } from '../types'

// TODO support tsconfig paths,
// TODO make any non js module external
// use the right resolver in esbuild
// maybe can add support for vue and other stuff by loading a vue file to a js file? i will let people pas the loaders for esbuild directly, this way they can pass them if they want
// mark as external modules that make stopTraversing true

type Args = {
    entryPoints: string[]
    esbuildOptions?: Partial<BuildOptions>
    resolver?: (cwd: string, id: string) => string
    stopTraversing?: (resolvedPath: string) => boolean
}

// resolver = defaultResolver,
// onEntry,
// stopTraversing,
// ignore = [],
// readFile = defaultRead,
export async function traverseWithEsbuild({
    entryPoints,
    resolver = defaultResolver,
    esbuildOptions = { plugins: [] },
    stopTraversing,
}: Args): Promise<TraversalResultType[]> {
    const destLoc = path.resolve(
        await fsp.mkdtemp(path.join(os.tmpdir(), 'dest')),
    )

    entryPoints = entryPoints.map((x) => path.resolve(x))

    try {
        const metafile = path.join(destLoc, 'meta.json')

        const esbuildCwd = process.cwd()
        await build(
            deepmerge(
                {
                    // splitting: true, // needed to dedupe modules
                    // external: externalPackages,

                    minifyIdentifiers: false,
                    minifySyntax: false,
                    minifyWhitespace: false,
                    mainFields: ['module', 'browser', 'main'],
                    sourcemap: false,
                    define: {
                        'process.env.NODE_ENV': JSON.stringify('dev'),
                        global: 'window',
                        // ...generateEnvReplacements(env),
                    },
                    inject: [
                        // require.resolve(
                        //     '@esbuild-plugins/node-globals-polyfill/process.js',
                        // ),
                    ],
                    // tsconfig: ,
                    loader: {
                        '.js': 'jsx',
                    },
                    plugins: [
                        ExternalButInMetafile(),
                        // NodeModulesPolyfillPlugin({ fs: true, crypto: true }), // TODO enable if in browser?
                        NodeResolvePlugin({
                            external: function external(resolved) {
                                if (
                                    stopTraversing &&
                                    stopTraversing(resolved)
                                ) {
                                    return {
                                        namespace: externalNamespace,
                                        path: resolved,
                                    }
                                }
                                return false
                            },
                            onUnresolved: () => {
                                return {
                                    external: true,
                                }
                            },
                            resolveOptions: {
                                preserveSymlinks: isRunningWithYarnPnp || false,
                                extensions: [...JS_EXTENSIONS],
                            },
                        }),
                    ].filter(Boolean),
                    bundle: true,
                    platform: 'node',
                    format: 'esm',
                    write: true,
                    entryPoints,
                    outdir: destLoc,
                    minify: false,
                    logLevel: 'info',
                    metafile,
                } as BuildOptions,
                esbuildOptions,
            ),
        )

        const meta: Metadata = JSON.parse(
            await (await fsp.readFile(metafile)).toString(),
        )
        // console.log(await (await fsp.readFile(metafile)).toString())

        const res = flatten(
            entryPoints.map((entry) => {
                return metaToTraversalResult({ meta, entry, esbuildCwd })
            }),
        ).map((x) => {
            return {
                ...x,
                resolvedImportPath: removeColon(x.resolvedImportPath),
                importer: removeColon(x.importer),
            }
        })
        return res
    } finally {
        await fsx.remove(destLoc)
    }
}

const externalNamespace = 'external-but-keep-in-metafile'
function ExternalButInMetafile(): Plugin {
    return {
        name: externalNamespace,
        setup(build) {
            const externalModule = 'externalModuleXXX'
            build.onResolve({ filter: new RegExp(externalModule) }, (args) => {
                if (args.path !== externalModule) {
                    return
                }
                return {
                    external: true,
                }
            })
            build.onLoad(
                {
                    filter: /.*/,
                    namespace: 'external-but-keep-in-metafile',
                },
                (args) => {
                    const contents = `export * from '${externalModule}'`
                    return { contents, loader: 'js' }
                },
            )
        },
    }
}

function removeColon(input: string) {
    const index = input.indexOf(':')
    if (index === -1) {
        return input
    }
    const clean = input.slice(index + 1)
    return clean
}

// TODO i cannot get the import paths with esbuild, output will only be with resolvedPath and importer

export function metaToTraversalResult({
    meta,
    entry,
    esbuildCwd = '.',
}: {
    meta: Metadata
    esbuildCwd: string
    entry: string
}): TraversalResultType[] {
    if (!path.isAbsolute(esbuildCwd)) {
        throw new Error('esbuildCwd must be an absolute path')
    }
    if (!path.isAbsolute(entry)) {
        throw new Error('entry must be an absolute path')
    }
    const alreadyProcessed = new Set<string>()
    let toProcess = [slash(path.relative(esbuildCwd, entry))]
    let result: TraversalResultType[] = []
    while (toProcess.length) {
        const newImports = flatten(
            toProcess.map((newEntry) => {
                if (alreadyProcessed.has(newEntry)) {
                    return
                }
                alreadyProcessed.add(newEntry)
                // newEntry = path.posix.normalize(newEntry) // TODO does esbuild always use posix?
                const input = meta.inputs[newEntry]
                if (!input) {
                    throw new Error(
                        `entry ${newEntry} is not present in esbuild metafile`,
                    )
                }
                const currentImports = input.imports.map((x) => x.path)
                // newImports.push(...currentImports)
                result.push(
                    ...currentImports.map(
                        (x): TraversalResultType => {
                            return {
                                importPath: '',
                                importer: path.join(esbuildCwd, newEntry),
                                resolvedImportPath: path.join(esbuildCwd, x),
                            }
                        },
                    ),
                )
                return currentImports
            }),
        ).filter(Boolean)
        toProcess = newImports
    }
    return unique(result, (x) => x.importPath + x.resolvedImportPath)
    // find the right output getting the key of the right output.inputs == input
    // get the imports of the inputs.[entry].imports and attach them the importer
    // do the same with the imports just found
    // return the list of input files
}
