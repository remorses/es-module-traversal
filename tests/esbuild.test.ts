import { Metadata } from 'esbuild'
import path from 'path'
import {
    metaToTraversalResult,
    traverseWithEsbuild,
} from '../src/traverseEsbuild'
import { osAgnosticResult } from './support'

it('metaToTraversalResult', async () => {
    const meta: Metadata = {
        outputs: {},
        inputs: {
            entry: {
                bytes: 0,
                imports: [{ path: 'some-file' }],
            },
            'some-file': {
                bytes: 0,
                imports: [{ path: 'other-file' }],
            },
            'other-file': {
                bytes: 0,
                imports: [],
            },
        },
    }
    const res = metaToTraversalResult({
        meta: meta as any,
        entry: '/usr/someFolder/entry',
        esbuildCwd: '/usr/someFolder',
    }).map(osAgnosticResult)
    // console.log(res)
    expect(res).toMatchSnapshot('simple metaToTraversalResult')
})
it('traverseWithEsbuild', async () => {
    const currentFile = path.resolve(__dirname, __filename)
    let res = await traverseWithEsbuild({
        entryPoints: [currentFile],
        esbuildOptions: {
            platform: 'node',
        },
    })
    res = res.map(osAgnosticResult)
    console.log(res)
    expect(res).toMatchSnapshot('traverseWithEsbuild')
})
