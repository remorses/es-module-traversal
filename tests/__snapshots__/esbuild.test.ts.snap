// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`metaToTraversalResult: simple metaToTraversalResult 1`] = `
Array [
  Object {
    "importPath": "",
    "importer": "someFolder/some-file",
    "resolvedImportPath": "someFolder/other-file",
  },
  Object {
    "importPath": "",
    "importer": "someFolder/entry",
    "resolvedImportPath": "someFolder/some-file",
  },
]
`;

exports[`traverseWithEsbuild: traverseWithEsbuild 1`] = `
Array [
  Object {
    "importPath": "",
    "importer": "src/esbuild/traverseEsbuild.ts",
    "resolvedImportPath": "../../../.config/yarn/link/@esbuild-plugins/all/esm/index.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/koa/lib/request.js",
    "resolvedImportPath": "node_modules/accepts/index.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/koa-convert/node_modules/koa-compose/index.js",
    "resolvedImportPath": "node_modules/any-promise/index.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/any-promise/register.js",
    "resolvedImportPath": "node_modules/any-promise/loader.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/any-promise/index.js",
    "resolvedImportPath": "node_modules/any-promise/register.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/fs-extra/lib/mkdirs/make-dir.js",
    "resolvedImportPath": "node_modules/at-least-node/index.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/koa/lib/response.js",
    "resolvedImportPath": "node_modules/cache-content-type/index.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/koa-convert/index.js",
    "resolvedImportPath": "node_modules/co/index.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/koa/lib/response.js",
    "resolvedImportPath": "node_modules/content-disposition/index.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/content-disposition/index.js",
    "resolvedImportPath": "node_modules/content-disposition/node_modules/safe-buffer/index.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/koa/lib/request.js",
    "resolvedImportPath": "node_modules/content-type/index.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/koa/lib/context.js",
    "resolvedImportPath": "node_modules/cookies/index.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/cookies/index.js",
    "resolvedImportPath": "node_modules/cookies/node_modules/depd/lib/browser/index.js",
  },
  Object {
    "importPath": "",
    "importer": "src/support.ts",
    "resolvedImportPath": "node_modules/debug/src/browser.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/debug/src/browser.js",
    "resolvedImportPath": "node_modules/debug/src/common.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/http-assert/index.js",
    "resolvedImportPath": "node_modules/deep-equal/index.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/deep-equal/index.js",
    "resolvedImportPath": "node_modules/deep-equal/lib/is_arguments.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/deep-equal/index.js",
    "resolvedImportPath": "node_modules/deep-equal/lib/keys.js",
  },
  Object {
    "importPath": "",
    "importer": "src/esbuild/traverseEsbuild.ts",
    "resolvedImportPath": "node_modules/deepmerge/dist/cjs.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/koa/lib/context.js",
    "resolvedImportPath": "node_modules/delegates/index.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/koa/lib/application.js",
    "resolvedImportPath": "node_modules/depd/lib/browser/index.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/koa/lib/response.js",
    "resolvedImportPath": "node_modules/destroy/index.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/on-finished/index.js",
    "resolvedImportPath": "node_modules/ee-first/index.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/koa/lib/response.js",
    "resolvedImportPath": "node_modules/encodeurl/index.js",
  },
  Object {
    "importPath": "",
    "importer": "src/esbuild/traverseEsbuild.ts",
    "resolvedImportPath": "node_modules/esbuild/lib/main.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/koa/lib/response.js",
    "resolvedImportPath": "node_modules/escape-html/index.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/koa/lib/request.js",
    "resolvedImportPath": "node_modules/fresh/index.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/fs-extra/lib/copy-sync/index.js",
    "resolvedImportPath": "node_modules/fs-extra/lib/copy-sync/copy-sync.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/fs-extra/lib/index.js",
    "resolvedImportPath": "node_modules/fs-extra/lib/copy-sync/index.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/fs-extra/lib/copy/index.js",
    "resolvedImportPath": "node_modules/fs-extra/lib/copy/copy.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/fs-extra/lib/index.js",
    "resolvedImportPath": "node_modules/fs-extra/lib/copy/index.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/fs-extra/lib/index.js",
    "resolvedImportPath": "node_modules/fs-extra/lib/empty/index.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/fs-extra/lib/ensure/index.js",
    "resolvedImportPath": "node_modules/fs-extra/lib/ensure/file.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/fs-extra/lib/index.js",
    "resolvedImportPath": "node_modules/fs-extra/lib/ensure/index.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/fs-extra/lib/ensure/index.js",
    "resolvedImportPath": "node_modules/fs-extra/lib/ensure/link.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/fs-extra/lib/ensure/symlink.js",
    "resolvedImportPath": "node_modules/fs-extra/lib/ensure/symlink-paths.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/fs-extra/lib/ensure/symlink.js",
    "resolvedImportPath": "node_modules/fs-extra/lib/ensure/symlink-type.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/fs-extra/lib/ensure/index.js",
    "resolvedImportPath": "node_modules/fs-extra/lib/ensure/symlink.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/fs-extra/lib/index.js",
    "resolvedImportPath": "node_modules/fs-extra/lib/fs/index.js",
  },
  Object {
    "importPath": "",
    "importer": "src/esbuild/traverseEsbuild.ts",
    "resolvedImportPath": "node_modules/fs-extra/lib/index.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/fs-extra/lib/index.js",
    "resolvedImportPath": "node_modules/fs-extra/lib/json/index.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/fs-extra/lib/json/index.js",
    "resolvedImportPath": "node_modules/fs-extra/lib/json/jsonfile.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/fs-extra/lib/json/index.js",
    "resolvedImportPath": "node_modules/fs-extra/lib/json/output-json-sync.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/fs-extra/lib/json/index.js",
    "resolvedImportPath": "node_modules/fs-extra/lib/json/output-json.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/fs-extra/lib/index.js",
    "resolvedImportPath": "node_modules/fs-extra/lib/mkdirs/index.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/fs-extra/lib/mkdirs/index.js",
    "resolvedImportPath": "node_modules/fs-extra/lib/mkdirs/make-dir.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/fs-extra/lib/index.js",
    "resolvedImportPath": "node_modules/fs-extra/lib/move-sync/index.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/fs-extra/lib/move-sync/index.js",
    "resolvedImportPath": "node_modules/fs-extra/lib/move-sync/move-sync.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/fs-extra/lib/index.js",
    "resolvedImportPath": "node_modules/fs-extra/lib/move/index.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/fs-extra/lib/move/index.js",
    "resolvedImportPath": "node_modules/fs-extra/lib/move/move.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/fs-extra/lib/index.js",
    "resolvedImportPath": "node_modules/fs-extra/lib/output/index.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/fs-extra/lib/index.js",
    "resolvedImportPath": "node_modules/fs-extra/lib/path-exists/index.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/fs-extra/lib/index.js",
    "resolvedImportPath": "node_modules/fs-extra/lib/remove/index.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/fs-extra/lib/remove/index.js",
    "resolvedImportPath": "node_modules/fs-extra/lib/remove/rimraf.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/fs-extra/lib/copy-sync/copy-sync.js",
    "resolvedImportPath": "node_modules/fs-extra/lib/util/stat.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/fs-extra/lib/copy-sync/copy-sync.js",
    "resolvedImportPath": "node_modules/fs-extra/lib/util/utimes.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/function-bind/index.js",
    "resolvedImportPath": "node_modules/function-bind/implementation.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/has/src/index.js",
    "resolvedImportPath": "node_modules/function-bind/index.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/graceful-fs/graceful-fs.js",
    "resolvedImportPath": "node_modules/graceful-fs/clone.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/fs-extra/lib/fs/index.js",
    "resolvedImportPath": "node_modules/graceful-fs/graceful-fs.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/graceful-fs/graceful-fs.js",
    "resolvedImportPath": "node_modules/graceful-fs/legacy-streams.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/graceful-fs/graceful-fs.js",
    "resolvedImportPath": "node_modules/graceful-fs/polyfills.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/is-core-module/index.js",
    "resolvedImportPath": "node_modules/has/src/index.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/koa/lib/context.js",
    "resolvedImportPath": "node_modules/http-assert/index.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/http-assert/index.js",
    "resolvedImportPath": "node_modules/http-assert/node_modules/http-errors/index.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/http-assert/node_modules/http-errors/index.js",
    "resolvedImportPath": "node_modules/http-assert/node_modules/setprototypeof/index.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/koa/lib/application.js",
    "resolvedImportPath": "node_modules/http-errors/index.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/http-errors/index.js",
    "resolvedImportPath": "node_modules/inherits/inherits_browser.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/is-core-module/index.js",
    "resolvedImportPath": "node_modules/is-core-module/core.json",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/resolve/lib/async.js",
    "resolvedImportPath": "node_modules/is-core-module/index.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/koa/lib/application.js",
    "resolvedImportPath": "node_modules/is-generator-function/index.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/fs-extra/lib/json/jsonfile.js",
    "resolvedImportPath": "node_modules/jsonfile/index.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/fs-extra/lib/json/output-json.js",
    "resolvedImportPath": "node_modules/jsonfile/utils.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/cookies/index.js",
    "resolvedImportPath": "node_modules/keygrip/index.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/koa/lib/application.js",
    "resolvedImportPath": "node_modules/koa-compose/index.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/koa/lib/application.js",
    "resolvedImportPath": "node_modules/koa-convert/index.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/koa-convert/index.js",
    "resolvedImportPath": "node_modules/koa-convert/node_modules/koa-compose/index.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/koa-static/index.js",
    "resolvedImportPath": "node_modules/koa-send/index.js",
  },
  Object {
    "importPath": "",
    "importer": "tests/support.ts",
    "resolvedImportPath": "node_modules/koa-static/index.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/koa-static/index.js",
    "resolvedImportPath": "node_modules/koa-static/node_modules/debug/src/browser.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/koa-static/node_modules/debug/src/browser.js",
    "resolvedImportPath": "node_modules/koa-static/node_modules/debug/src/common.js",
  },
  Object {
    "importPath": "",
    "importer": "tests/support.ts",
    "resolvedImportPath": "node_modules/koa/lib/application.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/koa/lib/application.js",
    "resolvedImportPath": "node_modules/koa/lib/context.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/koa/lib/application.js",
    "resolvedImportPath": "node_modules/koa/lib/request.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/koa/lib/application.js",
    "resolvedImportPath": "node_modules/koa/lib/response.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/koa/lib/application.js",
    "resolvedImportPath": "node_modules/koa/node_modules/debug/src/browser.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/koa/node_modules/debug/src/browser.js",
    "resolvedImportPath": "node_modules/koa/node_modules/debug/src/debug.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/koa/node_modules/debug/src/debug.js",
    "resolvedImportPath": "node_modules/koa/node_modules/ms/index.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/type-is/index.js",
    "resolvedImportPath": "node_modules/media-typer/index.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/mime-db/index.js",
    "resolvedImportPath": "node_modules/mime-db/db.json",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/mime-types/index.js",
    "resolvedImportPath": "node_modules/mime-db/index.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/cache-content-type/index.js",
    "resolvedImportPath": "node_modules/mime-types/index.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/koa-static/node_modules/debug/src/common.js",
    "resolvedImportPath": "node_modules/ms/index.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/accepts/index.js",
    "resolvedImportPath": "node_modules/negotiator/index.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/negotiator/index.js",
    "resolvedImportPath": "node_modules/negotiator/lib/charset.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/negotiator/index.js",
    "resolvedImportPath": "node_modules/negotiator/lib/encoding.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/negotiator/index.js",
    "resolvedImportPath": "node_modules/negotiator/lib/language.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/negotiator/index.js",
    "resolvedImportPath": "node_modules/negotiator/lib/mediaType.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/koa/lib/application.js",
    "resolvedImportPath": "node_modules/on-finished/index.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/koa/lib/application.js",
    "resolvedImportPath": "node_modules/only/index.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/koa/lib/request.js",
    "resolvedImportPath": "node_modules/parseurl/index.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/resolve-path/index.js",
    "resolvedImportPath": "node_modules/path-is-absolute/index.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/resolve/lib/node-modules-paths.js",
    "resolvedImportPath": "node_modules/path-parse/index.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/koa-send/index.js",
    "resolvedImportPath": "node_modules/resolve-path/index.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/resolve-path/index.js",
    "resolvedImportPath": "node_modules/resolve-path/node_modules/http-errors/index.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/resolve-path/node_modules/http-errors/index.js",
    "resolvedImportPath": "node_modules/resolve-path/node_modules/inherits/inherits_browser.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/resolve-path/node_modules/http-errors/index.js",
    "resolvedImportPath": "node_modules/resolve-path/node_modules/setprototypeof/index.js",
  },
  Object {
    "importPath": "",
    "importer": "src/support.ts",
    "resolvedImportPath": "node_modules/resolve/index.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/resolve/index.js",
    "resolvedImportPath": "node_modules/resolve/lib/async.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/resolve/lib/async.js",
    "resolvedImportPath": "node_modules/resolve/lib/caller.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/resolve/index.js",
    "resolvedImportPath": "node_modules/resolve/lib/core.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/resolve/lib/core.js",
    "resolvedImportPath": "node_modules/resolve/lib/core.json",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/resolve/index.js",
    "resolvedImportPath": "node_modules/resolve/lib/is-core.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/resolve/lib/async.js",
    "resolvedImportPath": "node_modules/resolve/lib/node-modules-paths.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/resolve/lib/async.js",
    "resolvedImportPath": "node_modules/resolve/lib/normalize-options.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/resolve/index.js",
    "resolvedImportPath": "node_modules/resolve/lib/sync.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/http-errors/index.js",
    "resolvedImportPath": "node_modules/setprototypeof/index.js",
  },
  Object {
    "importPath": "",
    "importer": "tests/support.ts",
    "resolvedImportPath": "node_modules/slash/index.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/statuses/index.js",
    "resolvedImportPath": "node_modules/statuses/codes.json",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/koa/lib/application.js",
    "resolvedImportPath": "node_modules/statuses/index.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/http-errors/index.js",
    "resolvedImportPath": "node_modules/toidentifier/index.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/keygrip/index.js",
    "resolvedImportPath": "node_modules/tsscmp/lib/index.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/koa/lib/response.js",
    "resolvedImportPath": "node_modules/type-is/index.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/fs-extra/lib/fs/index.js",
    "resolvedImportPath": "node_modules/universalify/index.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/koa/lib/response.js",
    "resolvedImportPath": "node_modules/vary/index.js",
  },
  Object {
    "importPath": "",
    "importer": "node_modules/cache-content-type/index.js",
    "resolvedImportPath": "node_modules/ylru/index.js",
  },
  Object {
    "importPath": "",
    "importer": "src/support.ts",
    "resolvedImportPath": "src/constants.ts",
  },
  Object {
    "importPath": "",
    "importer": "tests/esbuild.test.ts",
    "resolvedImportPath": "src/esbuild/index.ts",
  },
  Object {
    "importPath": "",
    "importer": "src/esbuild/index.ts",
    "resolvedImportPath": "src/esbuild/traverseEsbuild.ts",
  },
  Object {
    "importPath": "",
    "importer": "tests/support.ts",
    "resolvedImportPath": "src/support.ts",
  },
  Object {
    "importPath": "",
    "importer": "tests/esbuild.test.ts",
    "resolvedImportPath": "tests/support.ts",
  },
]
`;
